### DISCUSSION TITLE
Permutation of whole plots

### DISCUSSION THREAD

**Discussion by qutkat** [gavinsimpson/permute#36] [Q&A]
Hello there,

I have longitudinal data with 3 samples per individual ID. I would like to treat each ID as a plot and then permute the whole plots so I am keeping all observations from each individual together when I permute. 
I tried using the following code per the Details section in the documentation but I am getting an error message. 

perm = how(within = Within(type = "none"), plots = Plots(type=variables$PID), nperm = 9999)

The error I am getting is Error in match.arg(type) : 'arg' must be NULL or a character vector . I believe this is because type is expecting type = c("none","free","series","grid"), however int he details section it says to use type = foo in order to permute whole blocks. Not sure what I am doing wrong?

I am fairly new to R - could someone please assist?  It would be greatly appreciated!

Many thanks
Katherine

**Comment by gavinsimpson** [comment]
You want:

```r
how(within = Within(type = "none"),
    plots = Plots(strata = factor(variables$PID), type = "free"),
    nperm = 9999)
```

changing `"free"` to one of the other options if you need a specific type of restricted permutation design (regular time series/transect (`"series"`) or regular spatial grid (`"grid"`)).

**Reply by qutkat** [reply]
Hi Gavin, Thank you so much for getting back to me. I have tried the code you suggested and the adonis results look reasonable, but I am still a little concerned as when I ran shuffle to see what was happening it looks like the individual PIDs are all being randomly permuted within each week, whereas I need all PIDs from the same individual to be held together in the permutation. This is the code I am using - perhaps I am doing it incorrectly? 
Prior to this I set variables$PID as a factor. 

perm = how(within = Within(type = "none"), plots = Plots(strata=factor(variables$PID), type = "free"), nperm = 9999)

adonis2(arcsin ~ variables$d_rating + variables$Weight_kg + variables$BMI + variables$Age , permutations = perm, parallel = 7, na.rm=TRUE, by="margin") 

variables[shuffle(nrow(variables), control = perm),c("PID", "week")]

**Reply by gavinsimpson** [reply]
I can't reproduce this. In fact I can demonstrate that things work correctly in a reproducible setting:

```r
# setup
library("permute")

set.seed(1)
f1 <- gl(5, 10)            # in order
f2 <- f[sample(length(f))] # in random order
```

The easy setting, where everything is in order:

```r
set.seed(21)
h1 <- how(within = Within(type = "none"), plots = Plots(strata = f1, type = "free"))
s1 <- shuffle(f, control = h1)
s1
f[s1]
```

```
r$> s1                                                                          
 [1]  1  2  3  4  5  6  7  8  9 10 21 22 23 24 25 26 27 28 29 30 41 42 43 44 45
[26] 46 47 48 49 50 11 12 13 14 15 16 17 18 19 20 31 32 33 34 35 36 37 38 39 40
r$> f[s1]                                                                       
 [1] 1 1 1 1 1 1 1 1 1 1 3 3 3 3 3 3 3 3 3 3 5 5 5 5 5 5 5 5 5 5 2 2 2 2 2 2 2 2
[39] 2 2 4 4 4 4 4 4 4 4 4 4
Levels: 1 2 3 4 5
```

Here the permutation is:

* `1 -> 1`
* `2 -> 3`
* `3 -> 5`
* `4 -> 2`
* `5 -> 4`

And this continues to work with the randomly ordered `f2`

```r
set.seed(21)
h2 <- how(within = Within(type = "none"), plots = Plots(strata = f2, type = "free"))
s2 <- shuffle(f2, control = h2)
s2
f2
f2[s2]
```

```
r$> s2                                                                          
 [1]  1  7  3  8  6  2  5 10 15 11  4 12 13 14 17 19 18  9 26 22 16 27 23 24 20
[26] 21 25 38 37 39 42 40 33 44 47 49 28 29 31 34 43 30 35 32 45 46 41 36 50 48
r$> f2                                                                          
 [1] 1 4 1 4 3 5 2 2 4 3 5 1 1 1 2 4 3 5 2 4 5 2 1 1 5 3 3 4 3 3 4 3 1 4 4 4 5 2
[39] 5 5 3 2 5 2 1 1 2 5 2 3
Levels: 1 2 3 4 5
r$> f2[s2]                                                                      
 [1] 1 2 1 2 5 4 3 3 2 5 4 1 1 1 3 2 5 4 3 2 4 3 1 1 4 5 5 2 5 5 2 5 1 2 2 2 4 3
[39] 4 4 5 3 4 3 1 1 3 4 3 5
Levels: 1 2 3 4 5
```

although it is much harder to see, but if you compare the last two outputs, all `4`s became `2`s etc.

As for the ordering *within* the `PID` issue, I can't reproduce that either. He's another reproducible example building on the elements above. Here I define a variable `week` that is in order `1, 2, ..., 10` alongside the correctly ordered `f1` from earlier. Then I define a second data frame, `df2`, using the first, but scrambled in exactly the same way as `f1` was scrambled to create `f2` earlier.

```r
df1 <- data.frame(f = f1, week = rep(1:10, times = 5))
# scramble df1 the same way we scrambled f1 to make f2
set.seed(1)
df2 <- df1[sample(length(f1)), ]
split(df2$week, df2$f)
split(df2$week[s2], df2$f)
```

The last two lines there are splitting the scrambled `week` according to `f`, and the same but for the shuffled (permuted) ordering of the rows in `df2`, the scrambled data frame

```
r$> split(df2$week, df2$f)                                                      
$`1`
 [1]  4  1 10  7  9  3  6  8  2  5
$`2`
 [1]  4  8  5  2 10  9  3  6  1  7
$`3`
 [1]  3  1  5  6  7  9  2  4  8 10
$`4`
 [1]  9  4  3 10  6  1  2  5  7  8
$`5`
 [1]  3  1  7  8  9  5  2  6  4 10

r$> split(df2$week, df2$f[s2])                                                  
$`1`
 [1]  4  1 10  7  9  3  6  8  2  5
$`2`
 [1]  9  4  3 10  6  1  2  5  7  8
$`3`
 [1]  4  8  5  2 10  9  3  6  1  7
$`4`
 [1]  3  1  7  8  9  5  2  6  4 10
$`5`
 [1]  3  1  5  6  7  9  2  4  8 10
```

If you compare `$2` in the first output and `$3` in the second, you'll notice that the order of the weeks is the same. This is because `2` got swapped with `3` in this particular permutation. Compare `$5` in the first output with `$4`, again as per expectation as in this permutation `5` is swapped with `4`. The `$1` components are the same in both outputs because `1` was swapped with `1` in this permutation.

It's impossible to say what is wrong with your setting as your example is not reproducible by me. Check that you redefine your `how()` object after coercing it to be a factor; `how()` stores the factor defining the strata of the plots within the control object used to define permutations. So if you change the data or factor *after* defining the control object (`h` in your example) then things will get all out of whack. For example, I define `h1` and `h2` here because the factor indicating to which group each observation belongs changes when I scrambled the original factor `f1`. If you were to reuse `h1` you'd get the wrong permutation because the stored strata don't match the data once we scramble `f1 to make `f2`.

**Reply by qutkat** [reply]
Thank you for taking the time to provide these examples. I have investigated further and it seems that the results of adonis are dependant on how the PIDs are sorted in the file when they are submitted. If I sort by PID so all the PIDs for an individual are together, then I am getting a significant result, whereas if I leave the file sorted by week so the PIDs are not grouped together, I get a very different result which is insignificant.  The results do not vary for either of these settings when I change set.seed. Do you have any idea why this might be happening? I'm wondering if it is something to do with adonis?

**Reply by gavinsimpson** [reply]
You need to supply a *reproducible* example.

Also, don't build a model like this:

```r
adonis2(arcsin ~ variables$d_rating + variables$Weight_kg + variables$BMI + variables$Age, ...)
```

use the `data` argument:

```r
adonis2(arcsin ~ d_rating + Weight_kg + BMI + Age, data = variables, ...)
```
